src/pymor/reductors/aaa.py

```diff
@@ -0,0 +1,429 @@
+            continue
+
+        svs0 = [svs[k] for k in range_S if i_itpl[k]]
+        itpl_part0 = [itpl_part[k] for k in range_S if i_itpl[k]]
+
+        for j in itertools.product(*(itpl_part[k] for k in range_S if not i_itpl[k])):
+            l_j = list(j)
+            for ii in range(len(i_itpl)):
+                if i_itpl[ii]:
+                    l_j.insert(ii, slice(None))
+            samples0 = samples[tuple(l_j)]
+            T_mat = 1
+            for k in range_S:
+                if i_itpl[k]:
+                    T_new = np.eye(len(itpl_part[k]))
+                else:
+                    idx = np.where(np.array(itpl_part[k]) == l_j[k])[0][0]
+                    T_new = np.eye(1, len(itpl_part[k]), idx)
+                T_mat = np.kron(T_mat, T_new)
+            LL = nd_loewner(samples0, svs0, itpl_part0)
+            L = np.vstack([L, LL @ T_mat])
+    return L
+
+
+def make_bary_func(itpl_nodes, itpl_vals, coefs, removable_singularity_tol=1e-14):
+    r"""Return function handle for (multivariate) barycentric form.
+
+    The multivariate barycentric form for two variables is given by
+
+    .. math::
+        H(s,p) = \frac{\sum_{i=1}^{k}\sum_{j=1}^{q}\frac{\alpha_{ij}H(s_i,p_j)}{(s-s_i)(p-p_j)}}
+            {\sum_{i=1}^{k}\sum_{j=1}^{q}\frac{\alpha_{ij}}{(s-s_i)(p-p_j)}}
+
+    where for :math:`i=1,\ldots,k` and :math:`j=1,\ldots,q` we have that :math:`s_i` and :math:`p_j`
+    are interpolation nodes, :math:`H(s_i,p_j)` are interpolation values and :math:`\alpha_{ij}`
+    represent the barycentric coefficients. This implementation can also handle single-variable
+    barycentric forms as well as barycentric forms with more than two variables.
+
+    Parameters
+    ----------
+    itpl_nodes
+        Nested list such that `itpl_nodes[i]` contains interpolation nodes of the
+        `i`-th variable.
+    itpl_vals
+        Vector of interpolation values with `len(itpl_nodes[0])*...*len(itpl_nodes[-1])`
+        entries.
+    coefs
+        Vector of barycentric coefficients with `len(itpl_nodes[0])*...*len(itpl_nodes[-1])`
+        entries.
+    removable_singularity_tol
+        Tolerance for evaluating the barycentric function at a removable singularity
+        and performing pole cancellation.
+
+    Returns
+    -------
+    bary_func
+        (Multi-variate) rational function in barycentric form.
+    """
+    def bary_func(*args):
+        pd = 1
+        # this loop is for pole cancellation which occurs at interpolation nodes
+        for arg, itpl_node in zip(args, itpl_nodes):
+            d = arg - itpl_node
+            d_zero = d[np.abs(d) < removable_singularity_tol]
+            if len(d_zero) > 0:
+                d_min_idx = np.argmin(np.abs(d))
+                d = np.eye(1, len(d), d_min_idx)
+            else:
+                d = 1 / d
+            pd = np.kron(pd, d)
+        m = coefs.T * pd
+        num = np.tensordot(m, itpl_vals, axes=1)
+        denom = np.inner(coefs.T, pd)
```

Comment:
If `coefs` is a 1D NumPy array, then `.T` doesn't do anything. I guess this could be simplified to

```suggestion
        coefs_pd = coefs * pd
        num = np.inner(coefs_pd, itpl_vals)
        denom = np.sum(coefs_pd)
```

Alternatively, if `itpl_vals` and `coefs` can instead be NumPy arrays of shape `(len(itpl_nodes[0]), ..., len(itpl_nodes[-1]))`, then `pd` doesn't have to be constructed.

Reply:
This noticeably improved performance, so this was a nice find! Anyway, I am not sure what you mean by "`pd` doesn't have to be constructed". We can of course easily reshape `itpl_vals` and `coefs`.

Reply:
Ok, great. Regarding `pd`, I meant dividing `coefs` by something like `d[i * (np.newaxis,) + (slice(None),) + (num_vars - i - 1) * (np.newaxis,)]` in the above loop.